gateway/ln-gateway/src/lib.rs

```diff
@@ -1108,48 +1113,51 @@ impl Gateway {
             }
         };
 
+        let per_federation_routing_fees: HashMap<FederationId, RoutingFees> =
+            per_federation_routing_fees
+                .unwrap_or_default()
+                .into_iter()
+                .map(|(k, v)| (k, v.into()))
+                .collect();
+
         dbtx.insert_entry(&GatewayConfigurationKey, &gateway_config)
             .await;
+        self.update_federation_routing_fees(
+            &mut dbtx,
+            &gateway_config,
+            &per_federation_routing_fees,
+        )
+        .await?;
         dbtx.commit_tx().await;
 
-        self.update_federation_routing_fees(routing_fees, &gateway_config)
-            .await?;
+        if let Err(e) = Self::register_all_federations(self, &gateway_config).await {
+            warn!("{e:?}")
+        };
+
         info!("Set GatewayConfiguration successfully.");
 
         Ok(())
     }
 
-    /// Updates the routing fees for every federation configuration. Also
-    /// triggers a re-register of the gateway with all federations.
-    ///
-    /// TODO: Once per-federation fees are supported, this function should only
-    /// update a single federation's routing fees at a time.
+    /// Updates the routing fees for every federation configuration.
     async fn update_federation_routing_fees(
         &self,
-        routing_fees: Option<String>,
+        dbtx: &mut DatabaseTransaction<'_, Committable>,
         gateway_config: &GatewayConfiguration,
+        per_federation_routing_fees: &HashMap<FederationId, RoutingFees>,
     ) -> Result<()> {
-        if let Some(fees_str) = routing_fees {
-            let routing_fees = GatewayFee::from_str(fees_str.as_str())?.0;
-            let mut dbtx = self.gateway_db.begin_transaction().await;
-            let configs = dbtx
-                .find_by_prefix(&FederationIdKeyPrefix)
-                .await
-                .collect::<Vec<_>>()
-                .await;
-
-            for (id_key, mut fed_config) in configs {
-                fed_config.fees = routing_fees;
-                dbtx.insert_entry(&id_key, &fed_config).await;
-            }
-
-            dbtx.commit_tx().await;
-
-            if let Err(e) = Self::register_all_federations(self, gateway_config).await {
-                warn!("{e:?}")
-            };
+        let configs: Vec<(FederationIdKey, FederationConfig)> = dbtx
+            .find_by_prefix(&FederationIdKeyPrefix)
+            .await
+            .collect::<Vec<_>>()
+            .await;
+        for (id_key, mut fed_config) in configs {
```

Comment:
You shouldn't need to iterate through all the federations here. You already have the `federation_id`, you can just directly update that federation's config

Reply:
Now that we don't need to iterate through all federations, this function is likely not necessary at all and the logic can be inlined in `handle_set_configuration_msg`

Reply:
I think there is some confusion about the semantics here. This commands currently sets the fees for all federations, so if some federation is missing, it will go to default. I think this is good because we need some way to set federations to default. But of course if we need to change just one federation, perhaps we're gonna need another command.

Reply:
Good point @douglaz , I tried to preserve this behavior, so in my point of view, set-configuration will overwrite every state in federation config (at least this was the current behavior, correct me if I'm wrong). Maybe if we want to "patch" a fee for a existing federation, we need another command.

Reply:
This was not the previous behavior. The previous behavior only modified the configs that were `Some` in the `SetConfigurationPayload`. If the field was not set, it would use the previous set configuration. We should preserve these semantics imo

Reply:
I think we should leave it as is and the new default will be applicable only for new connected federations